File: hummingbird/androidApp/app/src/main/java/net/acumenus/hummingbird/data/PatientCommunicationModels.kt
package net.acumenus.hummingbird.data

import java.util.UUID

data class PatientCommunicationTopic(
    val code: String,
    val label: String,
)

data class PatientCommunicationUnit(
    val id: Int,
    val label: String,
)

data class PatientCommunicationPool(
    val poolUuid: String,
    val label: String,
)

data class PatientCommunicationMessage(
    val messageUuid: String,
    val senderDisplayRole: String,
    val visibility: String,
    val messageKind: String,
    val body: String?,
    val deliveryState: String,
    val sentAt: String?,
)

data class PatientCommunicationWorkItem(
    val workItemUuid: String,
    val threadUuid: String,
    val patientContextRef: String?,
    val topic: PatientCommunicationTopic,
    val unit: PatientCommunicationUnit?,
    val pool: PatientCommunicationPool,
    val status: String,
    val ownershipState: String,
    val assignedToMe: Boolean,
    val workItemVersion: Int,
    val threadVersion: Int,
    val lastMessageAt: String?,
    val dueAt: String?,
    val escalateAt: String?,
    val isResponseDue: Boolean,
    val isEscalationDue: Boolean,
    val closedAt: String?,
    val messages: List<PatientCommunicationMessage> = emptyList(),
    val hasEarlierMessages: Boolean = false,
)

data class PatientCommunicationInbox(
    val items: List<PatientCommunicationWorkItem>,
    val count: Int,
)

data class PatientCommunicationMutationResult(
    val workItem: PatientCommunicationWorkItem?,
    val message: PatientCommunicationMessage?,
    val eventUuid: String?,
    val replayed: Boolean,
)

data class PatientCommunicationRouteActions(
    val canRelease: Boolean,
    val canReassign: Boolean,
    val canReroute: Boolean,
)

data class PatientCommunicationRouteReason(
    val code: String,
    val label: String,
)

data class PatientCommunicationRouteReasonOptions(
    val release: List<PatientCommunicationRouteReason>,
    val reassign: List<PatientCommunicationRouteReason>,
    val reroute: List<PatientCommunicationRouteReason>,
)

data class PatientCommunicationReassignCandidate(
    val membershipUuid: String,
    val label: String,
    val membershipRole: String,
)

data class PatientCommunicationRerouteCandidate(
    val poolUuid: String,
    val label: String,
    val scopeType: String,
    val unit: PatientCommunicationUnit?,
)

data class PatientCommunicationRouteCandidates(
    val workItemUuid: String,
    val workItemVersion: Int,
    val threadVersion: Int,
    val actions: PatientCommunicationRouteActions,
    val reasonOptions: PatientCommunicationRouteReasonOptions,
    val reassignCandidates: List<PatientCommunicationReassignCandidate>,
    val rerouteCandidates: List<PatientCommunicationRerouteCandidate>,
)

enum class PatientCommunicationRoutingAction(
    val label: String,
) {
    Release("Release to team"),
    Reassign("Reassign responder"),
    Reroute("Reroute to another team"),
}

/**
 * Client-side routing policy is deliberately narrower than presentation state. The server's
 * action flags remain authoritative, while this layer fails closed on unknown codes and prevents
 * a stale or fabricated reason/opaque target from becoming a command.
 */
object PatientCommunicationRoutingPolicy {
    const val MAX_REASON_OPTIONS = 12
    const val MAX_CANDIDATES = 50
    const val MAX_LABEL_LENGTH = 120

    val releaseReasonCodes = setOf(
        "return_to_team",
        "shift_handoff",
        "responder_unavailable",
        "incorrect_assignment",
    )
    val reassignReasonCodes = setOf(
        "supervisor_assignment",
        "shift_handoff",
        "coverage_change",
        "workload_balance",
    )
    val rerouteReasonCodes = setOf(
        "wrong_team",
        "unit_transfer",
        "service_change",
        "specialty_needed",
    )
    val membershipRoles = setOf("responder", "triage", "supervisor")
    val scopeTypes = setOf("unit", "facility", "enterprise")

    fun canOpen(canRespond: Boolean, item: PatientCommunicationWorkItem): Boolean =
        canRespond && item.status == "open"

    fun matchesDisplayedItem(
        candidates: PatientCommunicationRouteCandidates,
        item: PatientCommunicationWorkItem,
    ): Boolean =
        candidates.workItemUuid == item.workItemUuid &&
            candidates.workItemVersion == item.workItemVersion &&
            candidates.threadVersion == item.threadVersion

    fun isAllowed(action: PatientCommunicationRoutingAction, actions: PatientCommunicationRouteActions): Boolean =
        when (action) {
            PatientCommunicationRoutingAction.Release -> actions.canRelease
            PatientCommunicationRoutingAction.Reassign -> actions.canReassign
            PatientCommunicationRoutingAction.Reroute -> actions.canReroute
        }

    fun reasons(
        action: PatientCommunicationRoutingAction,
        options: PatientCommunicationRouteReasonOptions,
    ): List<PatientCommunicationRouteReason> = when (action) {
        PatientCommunicationRoutingAction.Release -> options.release
        PatientCommunicationRoutingAction.Reassign -> options.reassign
        PatientCommunicationRoutingAction.Reroute -> options.reroute
    }

    fun allowedReasonCodes(action: PatientCommunicationRoutingAction): Set<String> = when (action) {
        PatientCommunicationRoutingAction.Release -> releaseReasonCodes
        PatientCommunicationRoutingAction.Reassign -> reassignReasonCodes
        PatientCommunicationRoutingAction.Reroute -> rerouteReasonCodes
    }

    fun canSubmit(
        candidates: PatientCommunicationRouteCandidates,
        action: PatientCommunicationRoutingAction?,
        reasonCode: String?,
        targetUuid: String?,
    ): Boolean {
        if (action == null || !isAllowed(action, candidates.actions)) return false
        if (reasons(action, candidates.reasonOptions).none { it.code == reasonCode }) return false
        return when (action) {
            PatientCommunicationRoutingAction.Release -> targetUuid == null
            PatientCommunicationRoutingAction.Reassign ->
                candidates.reassignCandidates.any { it.membershipUuid == targetUuid }
            PatientCommunicationRoutingAction.Reroute ->
                candidates.rerouteCandidates.any { it.poolUuid == targetUuid }
        }
    }
}

enum class PatientCommunicationCloseReason(
    val wireValue: String,
    val label: String,
) {
    QuestionAnswered("question_answered", "Question answered"),
    Duplicate("duplicate", "Duplicate conversation"),
    Transferred("transferred", "Transferred to another workflow"),
    PatientRequested("patient_requested", "Patient requested closure"),
    Other("other", "Other approved reason"),
}

/** Server-derived capability hints; every resource request remains independently gated. */
object PatientCommunicationAccess {
    fun isEligible(me: MeData?): Boolean = me?.canViewPatientCommunications == true

    fun canRespond(me: MeData?): Boolean =
        me?.canViewPatientCommunications == true && me.canRespondPatientCommunications
}

/**
 * PHI-minimized routing for the server-authorized attention item returned by `/for-you`.
 *
 * The type is retained across every local role queue. Navigation remains stricter: the item must
 * also have the communications domain and an exact, canonical work-item UUID identifier.
 */
object PatientCommunicationForYou {
    const val TYPE = "patient_communication"
    const val DOMAIN = "communications"
    const val ID_PREFIX = "patient-communication-"
    const val TITLE = "Patient message needs attention"
    const val SUBTITLE = "Open Messages to review this care-team communication."

    fun isType(type: String): Boolean = type == TYPE

    fun isRestrictedCandidate(id: String, type: String, domain: String?): Boolean =
        isType(type) || domain == DOMAIN || id.startsWith(ID_PREFIX)

    fun isExactRoutableTriple(id: String, type: String, domain: String?): Boolean =
        isType(type) && domain == DOMAIN && canonicalWorkItemUuid(id) != null

    fun isAttentionItem(item: ForYouItem): Boolean =
        isType(item.type) && item.domain == DOMAIN

    fun workItemUuid(item: ForYouItem): String? {
        if (!isAttentionItem(item)) return null
        return canonicalWorkItemUuid(item.id)
    }

    private fun canonicalWorkItemUuid(id: String): String? {
        if (!id.startsWith(ID_PREFIX)) return null
        val candidate = id.drop(ID_PREFIX.length)
        return runCatching { PatientCommunicationCommandIds.requireUuid(candidate) }.getOrNull()
    }

    fun urgencyLabel(tier: String): String = when (tier.lowercase()) {
        "critical" -> "Immediate attention"
        "warning" -> "Urgent"
        "success" -> "Within response target"
        else -> "Needs review"
    }
}

/** UUIDs are generated at the explicit user-action boundary and retained in memory for exact retry. */
object PatientCommunicationCommandIds {
    fun next(): String = UUID.randomUUID().toString()

    fun requireUuid(value: String): String {
        val normalized = UUID.fromString(value).toString()
        require(normalized == value) { "Idempotency keys must be canonical UUIDs." }
        return normalized
    }
}

enum class PatientCommunicationAttention {
    EscalationDue,
    ResponseDue,
    AwaitingResponse,
    Responded,
    Closed,
}

object PatientCommunicationPresentation {
    private val claimableOwnershipStates = setOf("pool_owned", "rerouted", "escalated")

    fun isClaimable(item: PatientCommunicationWorkItem): Boolean =
        item.status == "open" && !item.assignedToMe && item.ownershipState in claimableOwnershipStates

    fun attention(item: PatientCommunicationWorkItem): PatientCommunicationAttention = when {
        item.status == "closed" -> PatientCommunicationAttention.Closed
        item.isEscalationDue -> PatientCommunicationAttention.EscalationDue
        item.isResponseDue -> PatientCommunicationAttention.ResponseDue
        item.ownershipState == "responded" -> PatientCommunicationAttention.Responded
        else -> PatientCommunicationAttention.AwaitingResponse
    }

    fun attentionLabel(item: PatientCommunicationWorkItem): String = when (attention(item)) {
        PatientCommunicationAttention.EscalationDue -> "Escalation due now"
        PatientCommunicationAttention.ResponseDue -> "Response due now"
        PatientCommunicationAttention.AwaitingResponse -> "Within response target"
        PatientCommunicationAttention.Responded -> "Care-team response sent"
        PatientCommunicationAttention.Closed -> "Closed"
    }
}

enum class PatientCommunicationRecovery {
    Reauthenticate,
    RefetchWithoutResend,
    ExplicitExactRetryAvailable,
    DiscardCommand,
}

/** No branch ever authorizes an automatic mutation resend. */
object PatientCommunicationMutationRecoveryPolicy {
    fun after(statusCode: Int?): PatientCommunicationRecovery = when {
        statusCode == 401 -> PatientCommunicationRecovery.Reauthenticate
        statusCode == 409 -> PatientCommunicationRecovery.RefetchWithoutResend
        statusCode == null || statusCode in 500..599 -> PatientCommunicationRecovery.ExplicitExactRetryAvailable
        else -> PatientCommunicationRecovery.DiscardCommand
    }
}

File: hummingbird/iosApp/Hummingbird/Networking/PatientCommunicationModels.swift
import Foundation

// DTOs for accountable staff patient-communications routes in
// hummingbird-bff.v1.yaml. Routing-policy plumbing remains outside the primary
// work-item DTO. Its opaque pool UUID is retained only long enough to bind a
// successful reroute projection to the exact explicitly selected destination;
// it is never rendered, logged, or persisted.

struct PatientCommunicationInboxData: Decodable, Equatable {
    let items: [PatientCommunicationWorkItem]
    let count: Int
}

struct PatientCommunicationWorkItem: Decodable, Equatable, Identifiable {
    let workItemUuid: String
    let threadUuid: String
    let patientContextRef: String?
    let topic: PatientCommunicationTopic
    let unit: PatientCommunicationUnit?
    let pool: PatientCommunicationPool
    let status: String
    let ownershipState: String
    let assignedToMe: Bool
    let workItemVersion: Int
    let threadVersion: Int
    let lastMessageAt: String
    let dueAt: String
    let escalateAt: String
    let isResponseDue: Bool
    let isEscalationDue: Bool
    let closedAt: String?
    let messages: [PatientCommunicationMessage]?
    let hasEarlierMessages: Bool?

    var id: String { workItemUuid }
    var isOpen: Bool { status == "open" }
    var canClaim: Bool {
        isOpen && !assignedToMe && ["pool_owned", "rerouted", "escalated"].contains(ownershipState)
    }
    var canReply: Bool { isOpen && assignedToMe }
    var canClose: Bool { canReply && ownershipState == "responded" }
}

struct PatientCommunicationTopic: Decodable, Equatable {
    let code: String
    let label: String
}

struct PatientCommunicationUnit: Decodable, Equatable {
    let id: Int
    let label: String
}

struct PatientCommunicationPool: Decodable, Equatable {
    // Used only to bind a confirmed reroute projection to the exact selected
    // opaque target. It is never rendered, logged, or persisted.
    let poolUuid: String?
    let label: String

    init(poolUuid: String? = nil, label: String) {
        self.poolUuid = poolUuid
        self.label = label
    }
}

struct PatientCommunicationMessage: Decodable, Equatable, Identifiable {
    let messageUuid: String
    let senderDisplayRole: String
    let visibility: String
    let messageKind: String
    let body: String?
    let deliveryState: String
    let sentAt: String?

    var id: String { messageUuid }
    var isPatientVisible: Bool { visibility == "patient_visible" }
    var isFromPatient: Bool {
        senderDisplayRole == "Patient" || senderDisplayRole == "Representative"
    }
}

struct PatientCommunicationMutationData: Decodable, Equatable {
    // Normal mutations return a verified projection. A content-minimized exact
    // replay of an already-committed cross-pool reroute returns null because the
    // original actor may no longer be authorized to read the destination state.
    let workItem: PatientCommunicationWorkItem?
    let message: PatientCommunicationMessage?
    let eventUuid: String?
    let replayed: Bool
}

enum PatientCommunicationRoutingAction: String, CaseIterable, Identifiable {
    case release
    case reassign
    case reroute

    var id: String { rawValue }

    var label: String {
        switch self {
        case .release: return "Release to team"
        case .reassign: return "Reassign owner"
        case .reroute: return "Reroute team"
        }
    }

    fileprivate var allowedReasonCodes: Set<String> {
        switch self {
        case .release:
            return ["return_to_team", "shift_handoff", "responder_unavailable", "incorrect_assignment"]
        case .reassign:
            return ["supervisor_assignment", "shift_handoff", "coverage_change", "workload_balance"]
        case .reroute:
            return ["wrong_team", "unit_transfer", "service_change", "specialty_needed"]
        }
    }

    func allows(reasonCode: String) -> Bool {
        allowedReasonCodes.contains(reasonCode)
    }
}

struct PatientCommunicationRoutingActions: Decodable, Equatable {
    let canRelease: Bool
    let canReassign: Bool
    let canReroute: Bool

    func allows(_ action: PatientCommunicationRoutingAction) -> Bool {
        switch action {
        case .release: return canRelease
        case .reassign: return canReassign
        case .reroute: return canReroute
        }
    }
}

struct PatientCommunicationRoutingReasonOption: Equatable, Identifiable {
    let action: PatientCommunicationRoutingAction
    let code: String
    let label: String

    var id: String { "\(action.rawValue):\(code)" }
}

struct PatientCommunicationRoutingReasonOptions: Equatable {
    let release: [PatientCommunicationRoutingReasonOption]
    let reassign: [PatientCommunicationRoutingReasonOption]
    let reroute: [PatientCommunicationRoutingReasonOption]

    func options(for action: PatientCommunicationRoutingAction) -> [PatientCommunicationRoutingReasonOption] {
        switch action {
        case .release: return release
        case .reassign: return reassign
        case .reroute: return reroute
        }
    }
}

enum PatientCommunicationMembershipRole: String, Equatable {
    case responder
    case triage
    case supervisor

    var label: String {
        switch self {
        case .responder: return "Responder"
        case .triage: return "Triage"
        case .supervisor: return "Supervisor"
        }
    }
}

struct PatientCommunicationReassignCandidate: Equatable, Identifiable {
    let membershipUuid: String
    let label: String
    let membershipRole: PatientCommunicationMembershipRole

    var id: String { membershipUuid }
}

enum PatientCommunicationPoolScope: String, Equatable {
    case unit
    case facility
    case enterprise

    var label: String {
        switch self {
        case .unit: return "Unit team"
        case .facility: return "Facility team"
        case .enterprise: return "Enterprise team"
        }
    }
}

struct PatientCommunicationRoutingUnit: Equatable {
    // The numeric unit id is intentionally not retained. Candidate selection is
    // always by the opaque pool UUID; this bounded label is display copy only.
    let label: String
}

struct PatientCommunicationRerouteCandidate: Equatable, Identifiable {
    let poolUuid: String
    let label: String
    let scopeType: PatientCommunicationPoolScope
    let unit: PatientCommunicationRoutingUnit?

    var id: String { poolUuid }
}

/// A short-lived, server-authorized selector surface. Its custom decoder applies
/// strict bounds and allowlists before any option can become actionable.
struct PatientCommunicationRouteCandidatesData: Decodable, Equatable {
    let workItemUuid: String
    let workItemVersion: Int
    let threadVersion: Int
    let actions: PatientCommunicationRoutingActions
    let reasonOptions: PatientCommunicationRoutingReasonOptions
    let reassignCandidates: [PatientCommunicationReassignCandidate]
    let rerouteCandidates: [PatientCommunicationRerouteCandidate]

    private enum CodingKeys: String, CodingKey {
        case workItemUuid
        case workItemVersion
        case threadVersion
        case actions
        case reasonOptions
        case reassignCandidates
        case rerouteCandidates
    }

    private struct RawActions: Decodable {
        let canRelease: Bool
        let canReassign: Bool
        let canReroute: Bool

        private enum CodingKeys: String, CodingKey, CaseIterable {
            case canRelease
            case canReassign
            case canReroute
        }

        init(from decoder: Decoder) throws {
            try patientCommunicationRequireExactKeys(decoder, CodingKeys.allCases.map(\.rawValue))
            let container = try decoder.container(keyedBy: CodingKeys.self)
            canRelease = try container.decode(Bool.self, forKey: .canRelease)
            canReassign = try container.decode(Bool.self, forKey: .canReassign)
            canReroute = try container.decode(Bool.self, forKey: .canReroute)
        }
    }

    private struct RawReasonOptions: Decodable {
        let release: [RawReason]
        let reassign: [RawReason]
        let reroute: [RawReason]

        private enum CodingKeys: String, CodingKey, CaseIterable {
            case release
            case reassign
            case reroute
        }

        init(from decoder: Decoder) throws {
            try patientCommunicationRequireExactKeys(decoder, CodingKeys.allCases.map(\.rawValue))
            let container = try decoder.container(keyedBy: CodingKeys.self)
            release = try container.decode([RawReason].self, forKey: .release)
            reassign = try container.decode([RawReason].self, forKey: .reassign)
            reroute = try container.decode([RawReason].self, forKey: .reroute)
        }
    }

    private struct RawReason: Decodable {
        let code: String
        let label: String

        private enum CodingKeys: String, CodingKey, CaseIterable {
            case code
            case label
        }

        init(from decoder: Decoder) throws {
            try patientCommunicationRequireExactKeys(decoder, CodingKeys.allCases.map(\.rawValue))
            let container = try decoder.container(keyedBy: CodingKeys.self)
            code = try container.decode(String.self, forKey: .code)
            label = try container.decode(String.self, forKey: .label)
        }
    }

    private struct RawReassignCandidate: Decodable {
        let membershipUuid: String
        let label: String
        let membershipRole: String

        private enum CodingKeys: String, CodingKey, CaseIterable {
            case membershipUuid
            case label
            case membershipRole
        }

        init(from decoder: Decoder) throws {
            try patientCommunicationRequireExactKeys(decoder, CodingKeys.allCases.map(\.rawValue))
            let container = try decoder.container(keyedBy: CodingKeys.self)
            membershipUuid = try container.decode(String.self, forKey: .membershipUuid)
            label = try container.decode(String.self, forKey: .label)
            membershipRole = try container.decode(String.self, forKey: .membershipRole)
        }
    }

    private struct RawRerouteCandidate: Decodable {
        let poolUuid: String
        let label: String
        let scopeType: String
        let unit: RawUnit?

        private enum CodingKeys: String, CodingKey, CaseIterable {
            case poolUuid
            case label
            case scopeType
            case unit
        }

        init(from decoder: Decoder) throws {
            try patientCommunicationRequireExactKeys(decoder, CodingKeys.allCases.map(\.rawValue))
            let container = try decoder.container(keyedBy: CodingKeys.self)
            poolUuid = try container.decode(String.self, forKey: .poolUuid)
            label = try container.decode(String.self, forKey: .label)
            scopeType = try container.decode(String.self, forKey: .scopeType)
            unit = try container.decodeIfPresent(RawUnit.self, forKey: .unit)
        }
    }

    private struct RawUnit: Decodable {
        // Decode the governed shape but deliberately discard the numeric id.
        let id: Int
        let label: String

        private enum CodingKeys: String, CodingKey, CaseIterable {
            case id
            case label
        }

        init(from decoder: Decoder) throws {
            try patientCommunicationRequireExactKeys(decoder, CodingKeys.allCases.map(\.rawValue))
            let container = try decoder.container(keyedBy: CodingKeys.self)
            id = try container.decode(Int.self, forKey: .id)
            label = try container.decode(String.self, forKey: .label)
        }
    }

    init(from decoder: Decoder) throws {
        try patientCommunicationRequireExactKeys(decoder, [
            "workItemUuid",
            "workItemVersion",
            "threadVersion",
            "actions",
            "reasonOptions",
            "reassignCandidates",
            "rerouteCandidates",
        ])
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let rawWorkItemUUID = try container.decode(String.self, forKey: .workItemUuid)
        guard let canonicalWorkItemUUID = Self.canonicalUUID(rawWorkItemUUID) else {
            throw DecodingError.dataCorruptedError(
                forKey: .workItemUuid,
                in: container,
                debugDescription: "work_item_uuid must be a canonical lowercase UUID"
            )
        }

        let workItemVersion = try container.decode(Int.self, forKey: .workItemVersion)
        let threadVersion = try container.decode(Int.self, forKey: .threadVersion)
        guard workItemVersion > 0, threadVersion > 0 else {
            throw DecodingError.dataCorruptedError(
                forKey: .workItemVersion,
                in: container,
                debugDescription: "Routing versions must be positive"
            )
        }

        let rawActions = try container.decode(RawActions.self, forKey: .actions)
        let rawReasons = try container.decode(RawReasonOptions.self, forKey: .reasonOptions)
        let rawReassign = try container.decode([RawReassignCandidate].self, forKey: .reassignCandidates)
        let rawReroute = try container.decode([RawRerouteCandidate].self, forKey: .rerouteCandidates)

        guard rawReasons.release.count <= 12,
              rawReasons.reassign.count <= 12,
              rawReasons.reroute.count <= 12,
              rawReassign.count <= 50,
              rawReroute.count <= 50 else {
            throw DecodingError.dataCorruptedError(
                forKey: .reasonOptions,
                in: container,
                debugDescription: "Routing option bounds exceeded"
            )
        }

        let releaseReasons = try Self.validReasons(
            rawReasons.release,
            action: .release,
            codingPath: decoder.codingPath
        )
        let reassignReasons = try Self.validReasons(
            rawReasons.reassign,
            action: .reassign,
            codingPath: decoder.codingPath
        )
        let rerouteReasons = try Self.validReasons(
            rawReasons.reroute,
            action: .reroute,
            codingPath: decoder.codingPath
        )

        let reassignCandidates = try rawReassign.map { raw in
            guard let membershipUuid = Self.canonicalUUID(raw.membershipUuid),
                  let label = Self.boundedLabel(raw.label),
                  let role = PatientCommunicationMembershipRole(rawValue: raw.membershipRole) else {
                throw Self.corrupted("Invalid reassign candidate", codingPath: decoder.codingPath)
            }
            return PatientCommunicationReassignCandidate(
                membershipUuid: membershipUuid,
                label: label,
                membershipRole: role
            )
        }
        try Self.requireUnique(
            reassignCandidates,
            id: \PatientCommunicationReassignCandidate.membershipUuid,
            codingPath: decoder.codingPath
        )

        let rerouteCandidates = try rawReroute.map { raw in
            guard let poolUuid = Self.canonicalUUID(raw.poolUuid),
                  let label = Self.boundedLabel(raw.label),
                  let scope = PatientCommunicationPoolScope(rawValue: raw.scopeType) else {
                throw Self.corrupted("Invalid reroute candidate", codingPath: decoder.codingPath)
            }
            let unit: PatientCommunicationRoutingUnit?
            if let rawUnit = raw.unit {
                guard rawUnit.id > 0, let unitLabel = Self.boundedLabel(rawUnit.label) else {
                    throw Self.corrupted("Invalid reroute unit", codingPath: decoder.codingPath)
                }
                unit = PatientCommunicationRoutingUnit(label: unitLabel)
            } else {
                unit = nil
            }
            guard (scope == .unit && unit != nil) || (scope != .unit && unit == nil) else {
                throw Self.corrupted("Reroute scope and unit are inconsistent", codingPath: decoder.codingPath)
            }
            return PatientCommunicationRerouteCandidate(
                poolUuid: poolUuid,
                label: label,
                scopeType: scope,
                unit: unit
            )
        }
        try Self.requireUnique(
            rerouteCandidates,
            id: \PatientCommunicationRerouteCandidate.poolUuid,
            codingPath: decoder.codingPath
        )

        guard (!rawActions.canRelease || !releaseReasons.isEmpty),
              (!rawActions.canReassign || (!reassignReasons.isEmpty && !reassignCandidates.isEmpty)),
              (!rawActions.canReroute || (!rerouteReasons.isEmpty && !rerouteCandidates.isEmpty)),
              (rawActions.canReassign || reassignCandidates.isEmpty),
              (rawActions.canReroute || rerouteCandidates.isEmpty) else {
            throw Self.corrupted("Routing actions and candidate options are inconsistent", codingPath: decoder.codingPath)
        }

        self.workItemUuid = canonicalWorkItemUUID
        self.workItemVersion = workItemVersion
        self.threadVersion = threadVersion
        self.reasonOptions = PatientCommunicationRoutingReasonOptions(
            release: releaseReasons,
            reassign: reassignReasons,
            reroute: rerouteReasons
        )
        self.reassignCandidates = reassignCandidates
        self.rerouteCandidates = rerouteCandidates
        self.actions = PatientCommunicationRoutingActions(
            canRelease: rawActions.canRelease,
            canReassign: rawActions.canReassign,
            canReroute: rawActions.canReroute
        )
    }

    func matches(_ item: PatientCommunicationWorkItem) -> Bool {
        workItemUuid == item.workItemUuid
            && workItemVersion == item.workItemVersion
            && threadVersion == item.threadVersion
    }

    func reasons(for action: PatientCommunicationRoutingAction) -> [PatientCommunicationRoutingReasonOption] {
        reasonOptions.options(for: action)
    }

    func containsTarget(_ targetUUID: String?, for action: PatientCommunicationRoutingAction) -> Bool {
        switch action {
        case .release:
            return targetUUID == nil
        case .reassign:
            guard let targetUUID else { return false }
            return reassignCandidates.contains { $0.membershipUuid == targetUUID }
        case .reroute:
            guard let targetUUID else { return false }
            return rerouteCandidates.contains { $0.poolUuid == targetUUID }
        }
    }

    private static func validReasons(
        _ rawReasons: [RawReason],
        action: PatientCommunicationRoutingAction,
        codingPath: [CodingKey]
    ) throws -> [PatientCommunicationRoutingReasonOption] {
        let options = try rawReasons.map { raw -> PatientCommunicationRoutingReasonOption in
            guard action.allows(reasonCode: raw.code), let label = boundedLabel(raw.label) else {
                throw corrupted("Invalid or unknown \(action.rawValue) reason", codingPath: codingPath)
            }
            return PatientCommunicationRoutingReasonOption(action: action, code: raw.code, label: label)
        }
        try requireUnique(options, id: \PatientCommunicationRoutingReasonOption.code, codingPath: codingPath)
        return options
    }

    private static func canonicalUUID(_ raw: String) -> String? {
        let pattern = "^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$"
        guard raw.range(of: pattern, options: .regularExpression) != nil,
              let uuid = UUID(uuidString: raw),
              raw == uuid.uuidString.lowercased() else {
            return nil
        }
        return raw
    }

    private static func boundedLabel(_ raw: String) -> String? {
        let label = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !label.isEmpty,
              label.count <= 120,
              label.unicodeScalars.allSatisfy({ !CharacterSet.controlCharacters.contains($0) }) else {
            return nil
        }
        return label
    }

    private static func requireUnique<T, ID: Hashable>(
        _ values: [T],
        id: KeyPath<T, ID>,
        codingPath: [CodingKey]
    ) throws {
        var seen: Set<ID> = []
        guard values.allSatisfy({ seen.insert($0[keyPath: id]).inserted }) else {
            throw corrupted("Duplicate routing option", codingPath: codingPath)
        }
    }

    private static func corrupted(_ description: String, codingPath: [CodingKey]) -> DecodingError {
        DecodingError.dataCorrupted(.init(codingPath: codingPath, debugDescription: description))
    }
}

private struct PatientCommunicationStrictCodingKey: CodingKey {
    let stringValue: String
    let intValue: Int?

    init?(stringValue: String) {
        self.stringValue = stringValue
        intValue = nil
    }

    init?(intValue: Int) {
        stringValue = String(intValue)
        self.intValue = intValue
    }
}

private func patientCommunicationRequireExactKeys(
    _ decoder: Decoder,
    _ expected: [String]
) throws {
    let container = try decoder.container(keyedBy: PatientCommunicationStrictCodingKey.self)
    let actual = Set(container.allKeys.map(\.stringValue))
    let allowed = Set(expected)
    guard actual == allowed else {
        throw DecodingError.dataCorrupted(.init(
            codingPath: decoder.codingPath,
            debugDescription: "Unexpected or missing patient-communication routing fields"
        ))
    }
}

enum PatientCommunicationCloseReason: String, CaseIterable, Identifiable {
    case questionAnswered = "question_answered"
    case patientRequested = "patient_requested"
    case duplicate
    case transferred
    case other

    var id: String { rawValue }

    var label: String {
        switch self {
        case .questionAnswered: return "Question answered"
        case .patientRequested: return "Patient requested closure"
        case .duplicate: return "Duplicate thread"
        case .transferred: return "Transferred to another workflow"
        case .other: return "Other"
        }
    }
}

/// Testable seam for an in-memory-only communications surface. Implementations
/// must not persist responses or drafts and must never enqueue failed writes.
@MainActor
protocol PatientCommunicationsRepository {
    func patientCommunicationsInbox(bearer: String) async throws -> PatientCommunicationInboxData
    func patientCommunicationThread(workItemUUID: String, bearer: String) async throws -> PatientCommunicationWorkItem
    func claimPatientCommunication(
        workItemUUID: String,
        workItemVersion: Int,
        threadVersion: Int,
        idempotencyKey: UUID,
        bearer: String
    ) async throws -> PatientCommunicationMutationData
    func replyToPatientCommunication(
        workItemUUID: String,
        workItemVersion: Int,
        threadVersion: Int,
        message: String,
        clientMessageUUID: UUID,
        idempotencyKey: UUID,
        bearer: String
    ) async throws -> PatientCommunicationMutationData
    func closePatientCommunication(
        workItemUUID: String,
        workItemVersion: Int,
        threadVersion: Int,
        reasonCode: PatientCommunicationCloseReason,
        idempotencyKey: UUID,
        bearer: String
    ) async throws -> PatientCommunicationMutationData
    func patientCommunicationRouteCandidates(
        workItemUUID: String,
        bearer: String
    ) async throws -> PatientCommunicationRouteCandidatesData
    func releasePatientCommunication(
        workItemUUID: String,
        workItemVersion: Int,
        threadVersion: Int,
        reasonCode: String,
        idempotencyKey: UUID,
        bearer: String
    ) async throws -> PatientCommunicationMutationData
    func reassignPatientCommunication(
        workItemUUID: String,
        workItemVersion: Int,
        threadVersion: Int,
        targetMembershipUUID: String,
        reasonCode: String,
        idempotencyKey: UUID,
        bearer: String
    ) async throws -> PatientCommunicationMutationData
    func reroutePatientCommunication(
        workItemUUID: String,
        workItemVersion: Int,
        threadVersion: Int,
        targetPoolUUID: String,
        reasonCode: String,
        idempotencyKey: UUID,
        bearer: String
    ) async throws -> PatientCommunicationMutationData
}

extension APIClient: PatientCommunicationsRepository {}

/// Discoverability is based only on the effective capability emitted by /me.
/// The server remains authoritative and also requires a current responsibility-
/// pool membership for every resource read and mutation.
enum PatientCommunicationsEligibility {
    static func isEligible(_ me: MeData?) -> Bool {
        me?.can.viewPatientCommunications == true
    }
}
